mempool/src/shared_mempool/types.rs

```diff
@@ -250,13 +252,77 @@ impl Ord for BatchId {
     }
 }
 
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct TimelineId(pub Vec<u64>);
+
+impl TimelineId {
+    pub(crate) fn new(num_buckets: usize) -> Self {
+        Self(vec![0; num_buckets])
+    }
+
+    pub(crate) fn update(&mut self, batch_id: &MultiBatchId) {
+        if self.0.len() != batch_id.0.len() {
+            return;
+        }
+
+        let updated: Vec<_> = self
+            .0
+            .iter()
+            .zip(batch_id.0.iter())
+            .map(|(&cur, &(_start, end))| std::cmp::max(cur, end))
+            .collect();
+
+        self.0 = updated;
+    }
+}
+
+impl From<Vec<u64>> for TimelineId {
+    fn from(vector: Vec<u64>) -> Self {
+        Self(vector)
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct MultiBatchId(pub Vec<(u64, u64)>);
+
+impl MultiBatchId {
+    pub(crate) fn from_timeline_ids(old: &TimelineId, new: &TimelineId) -> Self {
+        Self(old.0.iter().cloned().zip(new.0.iter().cloned()).collect())
+    }
+}
+
+// Note: in rev order to check significant pairs first
+impl PartialOrd for MultiBatchId {
+    fn partial_cmp(&self, other: &MultiBatchId) -> Option<std::cmp::Ordering> {
+        for (&self_pair, &other_pair) in self.0.iter().rev().zip(other.0.iter().rev()) {
+            let ordering = self_pair.cmp(&other_pair);
+            if ordering != Ordering::Equal {
+                return Some(ordering);
+            }
+        }
+        Some(Ordering::Equal)
+    }
+}
+
+impl Ord for MultiBatchId {
```

Comment:
I thought we can derive Ord if PartialOrd is implemented or another way around

Reply:
Ah, looks like it's recommended to implement PartialOrd using Ord, which makes sense.
https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#how-can-i-implement-partialord